Sidebar: Reader Subscriptions

kean opened the pull request:
This PR adds the initial `ReaderSubscriptionsView` implementation. You can see you subscriptions, search, and select them. I'm planning to fix issues and the remaining features in the next PR.

<img width="1316" alt="Screenshot 2024-09-05 at 6 01 52 PM" src="https://github.com/user-attachments/assets/e17f6474-5e80-41dd-9014-d5e44cf9a348">


<img width="1316" alt="Screenshot 2024-09-05 at 5 49 16 PM" src="https://github.com/user-attachments/assets/937c98e8-9e59-4ff3-8a7a-035c7af9555f">

To test:

## Regression Notes
1. Potential unintended areas of impact


2. What I did to test those areas of impact (or what existing automated tests I relied on)


3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

kean committed:
Add ReaderSubscriptionsView

dangermattic commented:
<!--
  1 Error: PR requires at least one label...
  1 Warning: PR is not assigned to a milest...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">PR requires at least one label.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

crazytonyli commented on WordPress/Classes/ViewRelated/Reader/ReaderSubscriptionsView.swift:
```diff
@@ -0,0 +1,191 @@
+        .buttonStyle(.plain)
+        .popover(isPresented: $isAddingSubscription) {
+            ReaderAddSubscriptionView()
+        }
```
What do you think using an alert? If users accidentally tap any area outside of this popover, while they are typing feed url, the popover would disappear, wouldn't it? I don't think that'll happen on an alert.

kean replied:
Good point. I added "Cancel" and "Add" buttons and disable interactive dismiss if input is not empty, which is all according to [HIG](https://developer.apple.com/design/human-interface-guidelines/popovers#Best-practices). Reference: "New Event" in the Calendar app.

<img width="594" alt="Screenshot 2024-09-06 at 8 56 47 AM" src="https://github.com/user-attachments/assets/dab72608-5d28-4015-8a18-06a749198160">

crazytonyli commented on WordPress/Classes/ViewRelated/Reader/ReaderSubscriptionsView.swift:
```diff
@@ -0,0 +1,191 @@
+        }
+    }
+
+    private var buttonAddSubscription: some View {
```
What do you think putting this button at the navigation bar, which I think is a pretty common pattern?

kean replied:
I added it to the list with the intention to make it more discoverable that you can add any types of sites using RSS. I could also add it to the navigation bar, but it just require a bit more effort to re-wire the popover 😬

kean replied:
I refactored it to a separate `ReaderAddSubscriptionButton` and added it to the new empty state view and navigation bar.

crazytonyli commented on WordPress/Classes/ViewRelated/Reader/ReaderSubscriptionsView.swift:
```diff
@@ -0,0 +1,191 @@
+            }
+            HStack {
+                TextField("Link", text: $url, prompt: Text(verbatim: "https://www.site.com"))
+                    .focused($isFocused)
```
Setting `textContentType`, disabling auto-completion, and maybe other things that help to make typing url eaiser?

kean replied:
Done

crazytonyli commented on WordPress/Classes/ViewRelated/Reader/ReaderSubscriptionsView.swift:
```diff
@@ -0,0 +1,191 @@
+                    .foregroundStyle(.secondary)
+            }
+            HStack {
+                TextField("Link", text: $url, prompt: Text(verbatim: "https://www.site.com"))
```
Use `example.com`, which is used in other places of the app?

kean replied:
Good idea, done. I copied the link from wp.com, but it's, of course, an actual site. It also needs no scheme ("https").

kean committed:
Merge branch 'trunk' into task/sidebar-reader-manage-subscriptions

kean committed:
Update ReaderAddSubscriptionView based on feedback

kean committed:
Update search placement

kean committed:
Make subscriptions view refreshable

kean committed:
Make EmptyStateView adaptiev

kean committed:
Extract ReaderAddSubscriptionButton

kean committed:
Remove unused visibleSubscriptions

kean committed:
Update Reader tint color

kean committed:
Fix unfollow button and add Reader to SharedStrings

kean committed:
Add support for subscribing to sites

kean committed:
Add support for changing notification settings directly from Reader Subscriptions

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

kean commented:
**Updates:**

- Addressed feedback
- Implemented “Add Subscription” feature
- Added support for changing notification settings directly from the subscription list
- Minor UI fixes and improvements

crazytonyli commented on Modules/Sources/WordPressUI/Views/EmptyStateView.swift:
```diff
@@ -25,10 +26,11 @@ public struct EmptyStateView<Label: View, Description: View, Actions: View>: Vie
             actions()
         }
-        .frame(maxWidth: 300)
+        .frame(maxWidth: horizontalSizeClass == .compact ? 300 : 420)
```
This is not something that has to be implemented, but is it possible to constraint the width to be a certain percentage of its "parent view"? Is that a thing in SwiftUI?

kean replied:
You should be able to do it with a `GeometryReader`. I think it'll work for now. It makes it somewhat easier to design empty state views when there are only two variants.

crazytonyli commented on WordPress/Classes/ViewRelated/Reader/Subscriptions/ReaderSubscriptionsView.swift:
```diff
@@ -0,0 +1,139 @@
+        List {
+            if !isShowingSearchResuts {
+                ReaderSubscriptionAddButton(style: .expanded)
+                    .listRowSeparator(.hidden)
```
Now that there is an plus button on the navigation bar, what do you think removing this one at the top of the list, which feels redundant?

kean replied:
I think you are right. This extra large button is just fluff. I don't expect most people to add subscriptions manually – you find something in search, you add. I removed it.

crazytonyli commented on WordPress/Classes/ViewRelated/Reader/Subscriptions/ReaderSubscriptionsView.swift:
```diff
@@ -0,0 +1,139 @@
+        .swipeActions(edge: .trailing) {
+            Button(SharedStrings.Reader.unfollow, role: .destructive) {
+                ReaderSubscriptionHelper().unfollow(site)
+            }.tint(.red)
```
It's a bit strange that we have to explicitly set the color here. I thought this button would be red automatically because its role is destructive.

kean replied:
You do seem to have to set the tint manually for every swipe action.

crazytonyli replied:
Not in [this API doc's sample code](https://developer.apple.com/documentation/swiftui/view/swipeactions(edge:allowsfullswipe:content:)) though. I didn't run the sample code, so not sure if the API doc is lying.

kean replied:
Unfortunately, I don't know why it doesn't work. It probably has to do with `tint` being set by one of the container views.

crazytonyli commented on WordPress/Classes/ViewRelated/Reader/Subscriptions/ReaderSubscriptionsView.swift:
```diff
@@ -0,0 +1,139 @@
+        } else {
+            subscriptions[index]
+        }
+    }
```
It seems the two functions above are not used. Can they be deleted?

kean replied:
They are used in `.onDelete(perform: delete)` – this syntax make it a bit hard to find 🤔

crazytonyli replied:
That's the `delete(_:)` function below, right? The `delete(at:)` is not used there.

kean replied:
`delete(at:)` is used by `ForEach` and there `delete(_:)` by swipe actions.

crazytonyli replied:
Oh, I see it now. It's pretty strange that `delete` could be used to refer `delete(at:)`.

crazytonyli commented on WordPress/Classes/ViewRelated/Reader/Subscriptions/ReaderSubscriptionsViewModel.swift:
```diff
@@ -0,0 +1,50 @@
+                self?.refreshTask = nil
+                continuation.resume()
+            })
+        }
```
Nitpick: what do you think making the `withUnsafeContinuation` call return API result, and updating the `error` and `refreshTask` states in the context of the `async` function?

kean replied:
Yeah, it's not ideal. I'll keep stuff async-first in the future PRs.

kean committed:
Remove expanded add subscription button
